sdk/metrics/sdk.go

```diff
@@ -0,0 +1,386 @@
+	// the use of lock-free algorithms, there may be more than one
+	// `record` in existence at a time, although at most one can
+	// be referenced from the `SDK.current` map.
+	record struct {
+		// labels is the LabelSet passed by the user.
+		labels *labels
+
+		// descriptor describes the metric instrument.
+		descriptor *api.Descriptor
+
+		// refcount counts the number of active handles on
+		// referring to this record.  active handles prevent
+		// removing the record from the current map.
+		refcount int64
+
+		// collectedEpoch is the epoch number for which this
+		// record has been exported.  This is modified by the
+		// `Collect()` method.
+		collectedEpoch int64
+
+		// modifiedEpoch is the latest epoch number for which
+		// this record was updated.  Generally, if
+		// modifiedEpoch is less than collectedEpoch, this
+		// record is due for reclaimation.
+		modifiedEpoch int64
+
+		// reclaim is an atomic to control the start of reclaiming.
+		reclaim int64
+
+		// recorder implements the actual RecordOne() API,
+		// depending on the type of aggregation.
+		recorder export.MetricAggregator
+
+		// next contains the next pointer for both the primary
+		// and the reclaim lists.
+		next doublePtr
+	}
+
+	// singlePointer wraps an unsafe.Pointer and supports basic
+	// load(), store(), clear(), and swapNil() operations.
+	singlePtr struct {
+		ptr unsafe.Pointer
+	}
+
+	// doublePtr is used for the head and next links of two lists.
+	doublePtr struct {
+		primary singlePtr
+		reclaim singlePtr
+	}
+)
+
+// New constructs a new *SDK that implements intermediate storage for
+// metric instrument events, suporting configurable aggregation.
+//
+// The SDK does not start any background process to periodically
+// collect metrics.  It is the caller's responsibility to arrange for
+// periodic collection by calling the `Collect()` method.
+func New(exporter export.MetricBatcher) *SDK {
+	m := &SDK{
+		pool: sync.Pool{
+			New: func() interface{} {
+				return &bytes.Buffer{}
+			},
+		},
+		exporter: exporter,
+	}
+	m.empty.meter = m
+	return m
+}
+
+// DefineLabels returns a LabelSet corresponding to the arguments.
+// Labels are de-duplicated, with last-value-wins semantics.
+func (m *SDK) DefineLabels(_ context.Context, kvs ...core.KeyValue) api.LabelSet {
+	// Note: This computes a canonical encoding of the labels to
+	// use as a map key.  It happens to use the encoding used by
+	// statsd for labels, allowing an optimization for statsd
+	// exporters.  This could be made configurable in the
+	// constructor, to support the same optimization for different
+	// exporters.
+
+	// Check for empty set.
+	if len(kvs) == 0 {
+		return &m.empty
+	}
+
+	// Sort and de-duplicate.
+	sorted := sortedLabels(kvs)
+	sort.Stable(&sorted)
+	oi := 1
+	for i := 1; i < len(sorted); i++ {
+		if sorted[i-1].Key == sorted[i].Key {
+			sorted[oi-1] = sorted[i]
+			continue
+		}
+		sorted[oi] = sorted[i]
+		oi++
+	}
+	sorted = sorted[0:oi]
+
+	// Serialize.
```

Comment:
I wonder if this step could be forwarded to the exporter, so it can prepare the labels in its own way? For example, prometheus exporter would likely do some validation/transformations of the labels and its values, so it can be quickly passed to CounterVec/GaugeVec.